cms/check_rule_counts

```diff
@@ -38,273 +39,96 @@ from sqlalchemy import func
-
-            replicating_rules_older_than_gauge.labels(**prom_labels).set(result)
-            print(prom_labels, result)
-
-            probe_metrics.gauge(name='rules.no_of_rules.replicating.{name}.sum_locks_stuck_cnt').labels(name=a_name).set(result)
-
-            # Number of Stuck files
-
-            query = (session.query(func.sum(models.ReplicationRule.locks_stuck_cnt))
-                     .filter(models.ReplicationRule.state == RuleState.STUCK)
-                     .filter(models.ReplicationRule.created_at <= timeLimit))
-            result = query.scalar() or 0
-            print('rules.no_of_files.stuck.%s.sum_locks_stuck_cnt %s' % (a_name, result))
-
-            stuck_locks_older_than_gauge.labels(**prom_labels).set(result)
-            print(prom_labels, result)
-
-            probe_metrics.gauge(name='rules.no_of_files.stuck.{name}.sum_locks_stuck_cnt',
-                                documentation='Number of stuck files older than X').labels(name=a_name).set(result)
-
-            # Number of Replicating files
-
-            query = (session.query(func.sum(models.ReplicationRule.locks_replicating_cnt))
-                     .filter(models.ReplicationRule.state.in_([RuleState.STUCK, RuleState.REPLICATING]))
-                     .filter(models.ReplicationRule.created_at <= timeLimit))
-            result = query.scalar() or 0
-            print('rules.no_of_files.replicating.%s.sum_locks_replicating_cnt %s' % (a_name, result))
-
-            replicating_locks_older_than_gauge.labels(**prom_labels).set(result)
-            print(prom_labels, result)
-
-            (probe_metrics.gauge(name='rules.no_of_files.replicating.{name}.sum_locks_replicating_cnt',
-                                 documentation='Number of replicating files older than X')
-             .labels(name=a_name)
-             .set(result))
+        states = {'REPLICATING': RuleState.REPLICATING,
+                  'OK': RuleState.OK,
+                  'INJECT': RuleState.INJECT,
+                  'STUCK': RuleState.STUCK,
+                  'SUSPENDED': RuleState.SUSPENDED,
+                  'WAITING_APPROVAL': RuleState.WAITING_APPROVAL }
+
+        # Number of days which will be used to aggregate rules/locks
+        older_than_n_days = [1, 7, 21, 90, 180, 365]
+
+        with PrometheusPusher() as manager:
+
+            # Rule count per state
+            result = (session.query(models.ReplicationRule.state, func.count(models.ReplicationRule.state))
+                    .group_by(models.ReplicationRule.state)
+                    .with_hint(models.ReplicationRule, 'INDEX_FFS(rules RULES_PK)', 'oracle')
+                    .all())
+
+            for state, num in result:
+                manager.gauge(name='rule_count_per_state',
+                              documentation='Number of rules in a given state',
+                              labelnames=['state']).labels(state=str(state.name)).set(num)
```

Comment:
As long as label names match the keys in labels, they are not needed. Please simplify this further.

Reply:
You mean as follows?
```
manager.gauge(name='rule_count_per_state',
              documentation='Number of rules in a given state').labels(state=str(state.name)).set(num)
```

If so, this doesn't work:
```
# python3 check_rule_counts_haozturktest.py 
Traceback (most recent call last):
  File "/probes/cms/check_rule_counts_haozturktest.py", line 75, in <module>
    manager.gauge(name='rule_count_per_state',
  File "/usr/local/lib/python3.9/site-packages/rucio/core/monitor.py", line 196, in labels
    prom=self._prom.labels(**labelkwargs),
  File "/usr/local/lib/python3.9/site-packages/prometheus_client/metrics.py", line 181, in labels
    raise ValueError('No label names were set when constructing %s' % self)
ValueError: No label names were set when constructing gauge:rucio_probes_rule_count_per_state
```

Reply:
Ah, the difference is shown here: https://github.com/ericvaandering/probes/blob/406cba6df688b60d093708798b3688ef9c7d61e5/cms/check_rse_files_rules#L83 This way the label names are specified in the parameterized string. Want to try it that way? In this case the statsd metric is explicitly defined. (Of course we are getting rid of the statsd)

Reply:
I see your point. In my opinion, keeping the label names within in the metric name as a concatenated string makes the visualization part much more difficult because
- you need to do string parsing to introduce filtering whereas if you pass labels separately you can leverage built-in "Label filters" in Grafana. Secondly you need to cast numerical values back to their original types to achieve filtering.
- increases the number of metrics and makes the search harder. When we first needed to scan through probes, it was a pain scroll through tens of different metrics

See this test dashboard to see what I mean [LINK](https://monit-grafana.cern.ch/d/c778836f-9268-4964-a441-aaf8d5a6fc2d/hasan-rucio-probes-125-test?orgId=11&var-older_than_days=1&from=1706256000000&to=1706270399000&var-state=REPLICATING) (I didn't configure the filters probably. This is just a demo to prove my point)

This is not easily possible with what you suggest as far as I know. What do you think about this?

Reply:
Sorry, but I think you are confused. But that's OK, this is confusing!

If we do pure prometheus, then the Rucio prometheus code is smart enough to remove those label names from the metrics. Of course, you do see metrics which look like that but those are coming through statsd and the prometheus converter. Take a look at the code here: https://github.com/rucio/probes/pull/122/files 

Now look at the dashboard here: https://monit-grafana.cern.ch/d/c494efb0-1070-465e-b5c9-13cd0695fdf4/rucio-data-missing-at-rses?orgId=11&editPanel=82 You will see the data is coming in with simple metric names so it's fine. The labels have been removed from the metric name.

Now for sure if you search for a metric containing not_OK_rules you will find a lot with site names in them, but if you look carefully those are coming from the statsd-exporter job, so a 2nd path (which we will turn off). 

Ergo, no parsing needed to get what you want. Both the way you are doing it and the way I am produce the same prometheus metric. The way I'm doing it holds the statsd metric constant while we are still producing them.

Reply:
Ah okay, now I see, thanks for the detailed explanation. I updated and tested it and it works as you suggested and I expected. Does it look okay now?